mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,811 @@
+    config.arch.num_evaluation, remaining_updates = divmod(
+        config.system.num_updates, config.system.num_updates_per_eval
+    )
+    config.arch.num_evaluation += (
+        remaining_updates != 0
+    )  # Add an evaluation step if the num_updates is not a multiple of num_evaluation
+    steps_per_rollout = (
+        len(config.arch.executor_device_ids)
+        * config.arch.n_threads_per_executor
+        * config.system.rollout_length
+        * config.arch.num_envs
+        * config.system.num_updates_per_eval
+    )
+
+    # Logger setup
+    logger = MavaLogger(config)
+    cfg: Dict = OmegaConf.to_container(config, resolve=True)
+    cfg["arch"]["devices"] = jax.devices()
+    pprint(cfg)
+
+    # Set up checkpointer
+    save_checkpoint = config.logger.checkpointing.save_model
+    if save_checkpoint:
+        checkpointer = Checkpointer(
+            metadata=config,  # Save all config as metadata in the checkpoint
+            model_name=config.logger.system_name,
+            **config.logger.checkpointing.save_args,  # Checkpoint args
+        )
+
+    # Executor setup and launch.
+    unreplicated_params = flax.jax_utils.unreplicate(learner_state.params)
+    params_queues: List = []
+    rollout_queues: List = []
+
+    for _d_idx, d_id in enumerate(  # Loop through each executor device
+        config.arch.executor_device_ids
+    ):
+        # Replicate params per executor device
+        device_params = jax.device_put(unreplicated_params, devices[d_id])
+        # Loop through each executor thread
+        for _thread_id in range(config.arch.n_threads_per_executor):
+            seeds = np_rng.integers(np.iinfo(np.int64).max, size=config.arch.num_envs).tolist()
+            params_queues.append(queue.Queue(maxsize=1))
+            rollout_queues.append(queue.Queue(maxsize=1))
+            params_queues[-1].put(device_params)
+            threading.Thread(
+                target=rollout,
+                args=(
+                    jax.device_put(key, devices[d_id]),
+                    config,
+                    rollout_queues[-1],
+                    params_queues[-1],
+                    apply_fns,
+                    learner_devices,
+                    d_id,
+                    seeds,
+                ),
+            ).start()
+
+    # Run experiment for the total number of updates.
+    max_episode_return = jnp.float32(0.0)
+    best_params = None
+    for eval_step in range(config.arch.num_evaluation):
+        training_start_time = time.time()
+        learner_speeds = []
+        rollout_times = []
+
+        episode_metrics = []
+        train_metrics = []
+
+        # Full or partial last eval step.
+        num_updates_in_eval = (
+            remaining_updates
+            if eval_step == config.arch.num_evaluation - 1 and remaining_updates
+            else config.system.num_updates_per_eval
+        )
```

Comment:
This is quite confusing, but also without brackets I don't think it's doing what you want it to. Can you please explain this a bit better
```suggestion
        num_updates_in_eval = (
            remaining_updates
            if (eval_step == config.arch.num_evaluation - 1) and remaining_updates
            else config.system.num_updates_per_eval
        )
```

Reply:
I ended up dropping this since it just complicated the code, #1083 should address this.
If our num_updates is not a multiple of num_evaluations we add an evaluation step and run the remaining updates. 
This changes the number of updates in this eval to be the remaining updates if they are not 0.